### DISCUSSION TITLE
Verifying the function call sequence

### DISCUSSION THREAD

**Discussion by StanPlatinum** [smackers/smack#551] [Ideas]
In the context of security and C programs, one could tamper with the call stack. Guided by Prof. Rakamaric (https://github.com/boogie-org/boogie/issues/191), we @ya0guang @Gao-Chuan notice that model stack behaviors can be introduced in the Boogie files generated by SMACK. 
However, it may be a non-trivial work and hope we could do some help if this can be a future feature in SMACK’s memory safety check.

**Comment by 1phan** [comment]
For example, if we have a program like this:
```
#include <stdio.h>
#include <stdlib.h>

void __attribute__((stdcall))  fun(int k, int c, int d, int e, int f, int g, int h, int i, int j, int n, int l)
{
    l = 0xfff;
}



void  fun1(void (*pfun)(int), int a, char c)
{
    pfun(a);
}

int main()
{
    fun1(fun, 1, 'a');
    return 0;
}
```

```
gcc -m32 -g test.c -o test
```

There will be a Segmentation fault in x86 system.
The same thing will also happen on x64  as long as we make the parameter list long enough.

In this case, when eip return from fun() to fun1(), esp will point to main's stack frame:
Before:
![before](https://user-images.githubusercontent.com/22027527/76806060-cbda1f80-67b6-11ea-8ae0-43c78edae868.png)
After:
![after](https://user-images.githubusercontent.com/22027527/76806067-d0063d00-67b6-11ea-9f40-c35905b0b34e.png)

**Comment by shaobo-he** [comment]
@keram88 Any thoughts?

**Comment by ya0guang** [comment]
Another example:
```c
#include <stdlib.h>


int foo() {
    int i;
    int* pi;

    pi = &i;
    pi += 12;
    *pi = 0;

    return 0;

}

int main() {
    foo();
    return 0;   
}
```

I messed up the return address (on stack) in _foo_, so returning to _main_ will not be performed correctly and segmentation fault will be raised.

**Comment by shaobo-he** [comment]
> Another example:
> 
> ```c
> #include <stdlib.h>
> 
> 
> int foo() {
>     int i;
>     int* pi;
> 
>     pi = &i;
>     pi += 12;
>     *pi = 0;
> 
>     return 0;
> 
> }
> 
> int main() {
>     foo();
>     return 0;   
> }
> ```
> 
> I messed up the return address (on stack) in _foo_, so returning to _main_ will not be performed correctly and segmentation fault will be raised.

For this example, SMACK should be able to report an error with the flag `--memory-safety`.
